### DISCUSSION TITLE
Github workflow job not failing even when script is throwing error

### DISCUSSION THREAD

**Discussion by Supreetam09** [community/community#26710] [Actions]
<p>A very basic step in our project’s github workflow job is like:</p>
<pre><code class="lang-auto">      - name: Run Composer install
        run: |
          composer self-update --1
          composer global require hirak/prestissimo
          composer validate --no-check-all --ansi
          composer install --no-interaction || composer install --no-interaction || composer install --no-interaction
</code></pre>
<p>The composer.json has a post-install-cmd step which runs a script to clear the cache but it is throwing error like:</p>
<pre><code class="lang-auto">Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 255
!!  
!!   // Clearing the cache for the dev environment with debug                       
!!   // true                                                                        
!!  
!!  
!!  In environments.php line 19:
!!                                                                                 
!!    Attempted to load class "EnvironmentDetector" from namespace "Acquia\Blt\Ro  
!!    bo\Common".                                                                  
!!    Did you forget a "use" statement for another namespace?                      
!!                                                                                 
!!  
!!  cache:clear [--no-warmup] [--no-optional-warmers] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] &lt;command&gt;
!!  
!!  
Script @auto-scripts was called via post-install-cmd
</code></pre>
<p>I don’t care about the error since I know how to fix it but the point here is the build step did not fail even when error was thrown.<br>
How can I ensure the step fails which in turn fails the job in case any error occurs?<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/6/e/6e4b9aac829c7ae4e4d2aec4e047e5f60df33b53.png" data-download-href="/uploads/short-url/fJIvbuwMRvlb5eMU6799BosNgcP.png?dl=1" title="Screenshot 2022-03-07 at 7.54.25 PM" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/90027304/181099334-a7c7eb58-10f8-463a-aa45-c85bfc5cb9e9.png" alt="Screenshot 2022-03-07 at 7.54.25 PM" data-base62-sha1="fJIvbuwMRvlb5eMU6799BosNgcP" width="690" height="332" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/6/e/6e4b9aac829c7ae4e4d2aec4e047e5f60df33b53_2_690x332.png, https://aws1.discourse-cdn.com/github/optimized/3X/6/e/6e4b9aac829c7ae4e4d2aec4e047e5f60df33b53_2_1035x498.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/6/e/6e4b9aac829c7ae4e4d2aec4e047e5f60df33b53_2_1380x664.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/6/e/6e4b9aac829c7ae4e4d2aec4e047e5f60df33b53_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Screenshot 2022-03-07 at 7.54.25 PM</span><span class="informations">3536×1706 294 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by airtower-luna** [comment]
<p>In general: A run step fails if there’s a non-zero exit code that isn’t handled (e.g. by a <code>||</code> or <code>if</code> or whatever). I don’t know under which conditions composer returns a non-zero exit code, but I do notice that the last line of your script is essentially “try <code>composer install --no-interaction</code> up to three times”. That seems fishy.</p>

**Comment by Supreetam09** [comment, marked as answer]
<p>This is solved now. Its not an issue with Github Actions.<br>
The problem I found is the main <code>composer.json</code> file was installing another two application using a bash script in <code>post-install-cmd</code>. Inside that bash script, two separate <code>composer install</code> command was written. The first <code>composer install</code> command was actually throwing the error while the second <code>composer install</code> inside that script was running successfully. I had to use <code>Set -e</code> inside bash script which causes the script to exit immediately if a command returns non-zero status. Since this was not set, even though the first command was failing but the second command was running successfully, Github Actions was not able to identify the error inside the script.</p>
<p>Solution: Whenever we are using any bash script, <code>Set -e</code> to exit immediately on a command failure.</p>

**Reply by iamutkarshtiwari** [reply]
Thank you for being awesome enough to post solution to your own question. This really helped!

**Reply by Marcono1234** [reply]
If you explicitly specify `shell: bash` it should also by default fail fast on errors (see [documentation](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference)).

**Reply by KuthumiPepple** [reply]
True. Specifying `shell:bash` is the right answer if you also want your workflow to fail when any command in a pipeline fails.

**Comment by KuthumiPepple** [comment]
`set -euo pipefail` will also exit if any command in a pipeline fails or if you use an unset variable. Most times, you would want this behaviour.
